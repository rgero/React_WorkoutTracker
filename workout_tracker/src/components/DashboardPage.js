import { useEffect, useState } from "react";
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Navigate } from "react-router-dom";

const Dashboard = () => { 
    const [authenticated, setAuthenticated] = useState(false);
    const [hasLoaded, setLoaded] = useState(false)

    useEffect(() => {
        const checkForLogin = async () => {
                const token = await AsyncStorage.getItem('token');
                if (token)
                {
                    setAuthenticated(true);
                    
                }
                setLoaded(true);
        }
        checkForLogin().catch(console.error);
    }, []);

    if (hasLoaded)
    {
        if (!authenticated) {
            return <Navigate replace to="/login" />;
        } else {
            return (
            <div>
                    <p>Welcome to your Dashboard</p>
            </div>
            );
        }
    } else {
        return (
            <div>
                Loading...
            </div>
        )
    }
};

export default Dashboard;