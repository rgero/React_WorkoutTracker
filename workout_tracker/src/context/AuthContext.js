import AsyncStorage from '@react-native-async-storage/async-storage';

import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const authReducer = (state, action) => {
    switch (action.type)
    {
        case "addError":
            return { ...state, errorMessage: action.payload};
        case "storeToken":
            return { errorMessage: "", token: action.payload};
        case "clearErrorMessage":
            return {...state, errorMessage: ""}
        case "signout":
            return { token: null, errorMessage: ""}
        default:
            return state;
    }
};

const tryLocalSignin = dispatch => async () => {
    const token = await AsyncStorage.getItem('token');
    if (token)
    {
        dispatch({ type: 'storeToken', payload: token})
    }
}

const clearErrorMessage = dispatch => () => {
    dispatch( {type: 'clearErrorMessage' })
}

const signUp = dispatch => {
    return async ({ email, displayName, password }) => {
        try {
            const response = await trackerAPI.post('/signup', { email, displayName, password });
            
            // Storing the Token
            await AsyncStorage.setItem('token', response.data.token);
            dispatch({type: "storeToken", payload: response.data.token})
        } catch (err) {
            console.log(err.message);
            dispatch({ type: 'addError', payload: "Sign-up failed. Please try again"})
        }
    };
}

const signIn = (dispatch) => {
    return async ({ email, password }) => {
        try {
            const response = await trackerAPI.post('/signin', { email, password });
            
            // Storing the Token
            await AsyncStorage.setItem('token', response.data.token);
            dispatch({type: "storeToken", payload: response.data.token})
        } catch (err) {
            console.log(err.message);
            dispatch({ type: 'addError', payload: "Login failed. Please try again"})
        }
    };
}

const signOut = dispatch => async () => {
    try {
        await AsyncStorage.removeItem('token');
        dispatch({type: "signout"})
    } catch (err)
    {
        console.log(err);
    }
}

export const {Provider, Context } = createDataContext(authReducer, 
                                                      {clearErrorMessage, signUp, signIn, signOut, tryLocalSignin}, 
                                                      {
                                                        token: null,
                                                        errorMessage: ""
                                                      }
                                                    )