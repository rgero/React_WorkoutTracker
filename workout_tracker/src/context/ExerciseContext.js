import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const exerciseReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        default:
            return state;
    }
};

const fetchExercises = dispatch => async () => {
    const response = await trackerAPI.get('/exercises');
    dispatch({type: 'fetch', payload: response.data})
};

const fetchExerciseByID = dispatch => async (targetID) => {
    const response = await trackerAPI.get(`/exercises/${targetID}`);
    dispatch({type: 'fetch', payload: response.data})
};


export const {Provider, Context} = createDataContext(
    exerciseReducer,
    {fetchExercises, fetchExerciseByID},
    []
);