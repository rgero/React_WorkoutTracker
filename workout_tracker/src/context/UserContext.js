import AsyncStorage from '@react-native-async-storage/async-storage';

import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const userReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        default:
            return state;
    }
};

const getUser = dispatch => async () => {
    const token = await AsyncStorage.getItem('token');
    if (token)
    {
        const response = await trackerAPI.get('/users');
        dispatch({type: 'fetch', payload: response.data})
    }
}

export const {Provider, Context} = createDataContext(
    userReducer,
    {getUser},
    []
);