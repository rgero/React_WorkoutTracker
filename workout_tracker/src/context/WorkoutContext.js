import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const workoutReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        case "create":
            return state.concat(action.payload);
        case "edit":
            return state.map( (workout) => {
                if (workout._id === action._id) {
                    return {
                        ...workout,
                        ...action.update
                    }
                } else {
                    return workout;
                }
            })
        case "delete":
            return state.filter( (testWorkout) => {
                return action.id !== testWorkout._id;
            });
        default:
            return state;
    }
};

const fetchWorkouts = dispatch => async () => {
    const response = await trackerAPI.get('/workouts');
    dispatch({type: 'fetch', payload: response.data})
};

const createWorkout = dispatch => async (workout) => 
{
    const response = await trackerAPI.post('/workouts', workout)
    dispatch({type: "create", payload: response.data});
};

const updateWorkout = dispatch => async (workout) =>
{
    const response = await trackerAPI.put(`/workouts/${workout._id}`, workout);
    dispatch({type: "edit", payload: response.data});
}

const deleteWorkout = dispatch => async (targetID) => 
{
    const response = await trackerAPI.delete(`/workouts/${targetID}`)
    dispatch({type: "delete", id: targetID});
}

export const {Provider, Context} = createDataContext(
    workoutReducer,
    {fetchWorkouts, createWorkout, deleteWorkout, updateWorkout},
    []
);