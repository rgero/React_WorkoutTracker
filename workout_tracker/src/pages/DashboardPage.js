import { useEffect, useContext, useState } from "react";
import { Navigate } from "react-router-dom";
import { ViewWorkoutsPage } from "../components/workout/ViewWorkoutsPage";
import {Context as AuthContext} from '../context/AuthContext';

const DashboardPage = () => { 
    const {state, tryLocalSignin} = useContext(AuthContext);

    useEffect(()=> {
        tryLocalSignin();
    }, [tryLocalSignin])


    if (!state.token) {
        return <Navigate replace to="/login" />;
    } else {
        return (
            <>
                <ViewWorkoutsPage/>
            </>
        );
    }

};

export default DashboardPage;