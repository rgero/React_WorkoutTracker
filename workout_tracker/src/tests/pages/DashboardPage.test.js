import React from 'react';
import TestRenderer from 'react-test-renderer';
import reactRouterDom from 'react-router-dom';
import { act } from 'react-dom/test-utils';

import {Context as AuthContext} from '../../context/AuthContext';
import DashboardPage from '../../pages/DashboardPage';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
}));

jest.mock('../../pages/workouts/ViewWorkoutsPage', () => ({
    ViewWorkoutsPage: ()=> {
        return (<div>ViewWorkoutsPage Rendered Here</div>);
    }
}));

describe("Dashboard Tests", ()=> {

    test("Token Presented - matches mocked View Workout Page", async ()=> {

        const testContextValues = {
            state: { token: "asinaidnsdin"},
            tryLocalSignin: jest.fn()
        }
   
        let element;
        await act(async () => {
            element = new TestRenderer.create(
                <AuthContext.Provider value={testContextValues}>
                        <DashboardPage/>
                </AuthContext.Provider>
            );
        })

        expect(element.toJSON()).toMatchSnapshot();
    })
})