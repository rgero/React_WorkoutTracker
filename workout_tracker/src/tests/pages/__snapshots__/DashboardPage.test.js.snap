// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Tests Token Presented - matches mocked View Workout Page 1`] = `
<div>
  ViewWorkoutsPage Rendered Here
</div>
`;
